stetho/src/main/java/com/facebook/stetho/inspector/elements/android/window/WindowRootViewCompactV18Impl.java

```diff
@@ -0,0 +1,48 @@
+package com.facebook.stetho.inspector.elements.android.window;
+
+import android.support.annotation.NonNull;
+import android.view.View;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class WindowRootViewCompactV18Impl extends WindowRootViewCompat {
+
+	private Field mViewsField;
+	private Object mWindowManagerGlobal;
+
+	WindowRootViewCompactV18Impl() {
+		try {
+			Class wmClz = Class.forName("android.view.WindowManagerGlobal");
+			Method getInstanceMethod = wmClz.getDeclaredMethod("getInstance");
+			mWindowManagerGlobal = getInstanceMethod.invoke(wmClz);
+			mViewsField = wmClz.getDeclaredField("mViews");
+			mViewsField.setAccessible(true);
+		} catch (ClassNotFoundException e) {
+			throw new RuntimeException("unfortunately, you cannot view the view tree of the dialog etc.", e);
+		} catch (NoSuchMethodException e) {
+			throw new RuntimeException("unfortunately, you cannot view the view tree of the dialog etc.", e);
+		} catch (IllegalAccessException e) {
+			throw new RuntimeException("unfortunately, you cannot view the view tree of the dialog etc.", e);
+		} catch (InvocationTargetException e) {
+			throw new RuntimeException("unfortunately, you cannot view the view tree of the dialog etc.", e);
+		} catch (NoSuchFieldException e) {
+			throw new RuntimeException("unfortunately, you cannot view the view tree of the dialog etc.", e);
+		}
+
+	}
+
+	@NonNull
+	@Override
+	public List<View> getRootViews() {
+		try {
+			return Collections.unmodifiableList(Arrays.asList((View[]) mViewsField.get(mWindowManagerGlobal)));
```

Comment:
because before api 19,when add or remove window,WindowManager will create a new mViews array,and copy origin views to this new array.

Reply:
I don't think you understood my concern.  You're accessing the "root views" via WindowRootViewCompat.get(...).getRootViews(), but this will cause WindowRootViewCompat to be pinned in memory as a singleton.  If each call to getRootViews returns the same list as before, then for, say, API 19, the system will never be able to get a new list of root views after the first list is provided.

This should be corrected two-fold:

1. Each call to `getRootViews` should get the current list of root views.
2. The behaviour of each `...Impl` class should be updated to reflect this and ensure that it has this property to the best degree each `...Impl` can guarantee it.

Reply:
API19,When adding or removing windows, WindowManager adds or removes views directly from WindowManagerGlobal's private final ArrayList <View> mViews instead of creating a new arraylist each time.


example  api23 
```java
package android.view;

public final class WindowManagerGlobal {

	...

	private final ArrayList<View> mViews = new ArrayList<View>();

       public static WindowManagerGlobal getInstance() {
          ...
               return sDefaultWindowManager;
      
       }
	public void addView(View view, ViewGroup.LayoutParams params,
           Display display, Window parentWindow) {

           ...
           mViews.add(view);
           ...

    }

}

```

Reply:
In that case, you at least need to consistently make a copy of the array each time getRootViews is called in each impl.  Right now, at least v19 behaves differently by pretending to return an immutable list but in actuality it is edited under the hood without warning.

If you make sure a new copy is made in each call I think we're fine here.